src/graph/strongly-connected-components.md

```diff
@@ -4,151 +4,148 @@ tags:
 
-Thus, we built next **algorithm** for selecting strongly connected components:
+Thus, in summary, we discussed the following algorithm to find strongly connected components:
 
-1st step. Run sequence of depth first search of graph $G$ which will return vertices with increasing exit time $tout$, i.e. some list $order$.
+ - Step 1. Run a sequence of depth first searches on $G$, which will yield some list (e.g. `order`) of vertices, sorted on increasing exit time $t_\text{out}$.
 
-2nd step. Build transposed graph $G^T$. Run a series of depth (breadth) first searches in the order determined by list $order$ (to be exact in reverse order, i.e. in decreasing order of exit times). Every set of vertices, reached after the next search, will be the next strongly connected component.
+- Step 2. Build the transpose graph $G^T$, and run a series of depth first searches on the vertices in reverse order (i.e., in decreasing order of exit times). Each depth first search will yield one strongly connected component.
 
-Algorithm asymptotic is $O(n + m)$, because it is just two depth (breadth) first searches.
+- Step 3 (optional). Build the condensation graph.
 
-Finally, it is appropriate to mention [topological sort](topological-sort.md) here. First of all, step 1 of the algorithm represents reversed topological sort of graph $G$ (actually this is exactly what vertices' sort by exit time means). Secondly, the algorithm's scheme generates strongly connected components by decreasing order of their exit times, thus it generates components - vertices of condensation graph - in topological sort order.
+The runtime complexity of the algorithm is $O(n + m)$, because depth first search is performed twice. Building the condensation graph is also $O(n+m).$
+
+Finally, it is appropriate to mention [topological sort](topological-sort.md) here. In step 2, the algorithm finds strongly connected components in decreasing order of their exit times. Thus, it finds components - vertices of the condensation graph - in an order corresponding to a topological sort of the condensation graph.
 
 ## Implementation
 ```cpp
-vector<vector<int>> adj, adj_rev;
-vector<bool> used;
-vector<int> order, component;
+vector<vector<int>> adj, adj_rev; // adjacency lists of G and G^T
+vector<bool> visited;             // keeps track of which nodes are already visited
+vector<int> order;                // sorted list of G's vertices by exit time
+vector<int> component;            // stores one strongly connected component at a time
  
 void dfs1(int v) {
-    used[v] = true;
-
+    visited[v] = true;
     for (auto u : adj[v])
-        if (!used[u])
+        if (!visited[u])
             dfs1(u);
 
     order.push_back(v);
 }
  
 void dfs2(int v) {
-    used[v] = true;
+    visited[v] = true;
     component.push_back(v);
 
     for (auto u : adj_rev[v])
-        if (!used[u])
+        if (!visited[u])
             dfs2(u);
 }
  
 int main() {
```

Comment:
May be nice to turn it into a function that takes adjacency lists, so that we don't need to boilerplate input.

Reply:
That's also done. There is now also a (passing) test.